src/optimization/Optimizer.cpp

```diff
@@ -12,101 +12,36 @@
 #include "Combiner.h"
 #include "ControlFlow.h"
 #include "Eliminator.h"
-#include "Inliner.h"
-#include "LiteralValues.h"
-#include "MemoryAccess.h"
 #include "Reordering.h"
 #include "log.h"
 
 using namespace vc4c;
 using namespace vc4c::optimizations;
 
-OptimizationPass::OptimizationPass(const std::string& name, const Pass pass, const std::size_t index) :
-    name(name), index(index), pass(pass)
-{
-}
-
-bool OptimizationPass::operator<(const OptimizationPass& other) const
-{
-    return index < other.index;
-}
+OptimizationPass::OptimizationPass(const std::string& name, const Pass pass) : name(name), pass(pass) {}
 
 void OptimizationPass::operator()(const Module& module, Method& method, const Configuration& config) const
 {
     pass(module, method, config);
 }
 
-bool OptimizationPass::operator==(const OptimizationPass& other) const
-{
-    return name == other.name && index == other.index;
-}
-
-OptimizationStep::OptimizationStep(const std::string& name, const Step step, const std::size_t index) :
-    name(name), index(index), step(step)
-{
-}
-
-bool OptimizationStep::operator<(const OptimizationStep& other) const
-{
-    return index < other.index;
-}
+OptimizationStep::OptimizationStep(const std::string& name, const Step step) : name(name), step(step) {}
 
 InstructionWalker OptimizationStep::operator()(
     const Module& module, Method& method, InstructionWalker it, const Configuration& config) const
 {
     return step(module, method, it, config);
 }
 
-bool OptimizationStep::operator==(const OptimizationStep& other) const
-{
-    return name.compare(other.name) == 0 && index == other.index;
-}
-
-/*
- * Fail-fast for unresolved function-calls
- */
-static InstructionWalker checkMethodCalls(
-    const Module& module, Method& method, InstructionWalker it, const Configuration& config)
-{
-    if(it.has<intermediate::MethodCall>())
-    {
-        throw CompilationError(CompilationStep::OPTIMIZER, "There should be no more function-calls", it->to_string());
-    }
-    return it;
-}
-
-static const std::set<OptimizationStep> SINGLE_STEPS = {
-    // replaces all remaining returns with jumps to the end of the kernel-function
-    OptimizationStep("EliminateReturns", eliminateReturn, 0),
+static const std::vector<OptimizationStep> SINGLE_STEPS = {
     // combined successive branches to the same label (e.g. end of switch-case)
-    OptimizationStep("CombineDuplicateBranches", combineDuplicateBranches, 10),
+    OptimizationStep("CombineDuplicateBranches", combineDuplicateBranches),
```

Comment:
In previous implementation, the order of optimizations are controlled by integers (specified in initialization of `OptimizationStep`). Now, how to control it? I want to check the way.

Typical examples are optimization that works only if related options are specified. 
In this case, it okey to just check `config.flag` and return if not in the entrance of the optimization.

There is no practical case to inject optimization? The current implementation is enough?

Reply:
I removed the possibilities of adding optimization steps, because I don't see any use-cases for it.

Now, the optimization-order is given by their order in the vector. So optimizations depending on one another just need to be on the correct order in the vector.

If you object to this, we can roll it back, but I was thinking of rewriting the enabling of optimizations in #75

Reply:
No, I had no objection to the change. I have the same opinion.
I just wanted to check your opinion.